### PR TITLE
[SPARK-34421][SQL] Resolve temporary functions and views in views with CTEs

### PR DESCRIPTION
**Pull request by peter-toth**
### What changes were proposed in this pull request?
This PR:
- Fixes a bug that prevents analysis of:
  ```
  CREATE TEMPORARY VIEW temp_view AS WITH cte AS (SELECT temp_func(0)) SELECT * FROM cte;
  SELECT * FROM temp_view
  ```
  by throwing:
  ```
  Undefined function: 'temp_func'. This function is neither a registered temporary function nor a permanent function registered in the database 'default'.
  ```
- and doesn't report analysis error when it should:
  ```
  CREATE TEMPORARY VIEW temp_view AS SELECT 0;
  CREATE VIEW view_on_temp_view AS WITH cte AS (SELECT * FROM temp_view) SELECT * FROM cte
  ```
  by properly collecting temporary objects from VIEW definitions with CTEs.

- Minor refactor to make the affected code more readable.

### Why are the changes needed?
To fix a bug introduced with https://github.com/apache/spark/pull/30567

### Does this PR introduce _any_ user-facing change?
Yes, the query works again.

### How was this patch tested?
Added new UT + existing ones.

### CODE REVIEW

**Comment by peter-toth** [comment]
cc @linhongliu-db, @cloud-fan

**Comment by gatorsmile** [comment]
Thanks for the fix! Not sure whether @cloud-fan and @LinhongLiu will review it, because today is the lunar new year! 

CC @allisonwang-db @maryannxue

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala
Comment (allisonwang-db) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3964]: Can we also add an assertion for the error message here?

Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Sure, added in https://github.com/apache/spark/pull/31550/commits/1cd1cb34bf82248f42e6b1927c7b65c4c25944c8

**Comment by peter-toth** [comment]
cc @HyukjinKwon as this became a blocker of 3.1.1

**Comment by HyukjinKwon** [comment]
Thanks for pinging me @peter-toth. I will share it in the vote thread.

**Comment by HyukjinKwon** [comment]
@cloud-fan @LinhongLiu @allisonwang-db @maryannxue, it would be great if reviewing this is prioritized to unblock the release.

**Comment by HyukjinKwon** [comment]
ping @cloud-fan @LinhongLiu @allisonwang-db @maryannxue

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala
Comment (linhongliu-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala:560]: @peter-toth, thanks for fixing this.
any particular reason to replace `collect` to `flatMap`?

**Review by linhongliu-db — COMMENTED** [review]
LGTM with minor comments

**Comment by linhongliu-db** [comment]
@peter-toth, just read the description, do we have analysis error when creating the permanent view with CTE and refers temp objects before #30567

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3951]: Could you use `withUserDefinedFunction`?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3947]: Is this issue only related to a temp function case? How about a permanent one?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala:560]: I have the same comment, too. IMO the previous one looks okay.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3980]: hm, this message looks correct, so `v3.0.2` has the same issue?
```
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.0.2
      /_/
         
Using Scala version 2.12.10 (OpenJDK 64-Bit Server VM, Java 11.0.2)
Type in expressions to have them evaluated.
Type :help for more information.

scala> val tempViewName = "temp_view"
scala> val viewName = "view_on_temp_view"
scala> sql(s"CREATE TEMPORARY VIEW $tempViewName AS SELECT 0")
scala> sql(
     |   s"""
     |     |CREATE VIEW $viewName AS
     |     |WITH cte AS (
     |     |  SELECT * FROM $tempViewName
     |     |)
     |     |SELECT * FROM cte
     |     |""".stripMargin)

scala> spark.table(viewName).show()
+---+                                                                           
|  0|
+---+
|  0|
+---+
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala:560]: seems like it simplifies `collect(...).flatten` to `flatMap(...)`

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3947]: Same question

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3972]: So the bug happens in permanent view. Is it really a regression?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala
Comment (linhongliu-db) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3947]: for the temp view, we should verify the view capture temp objects correctly, so you should also add a permanent view/function, table with the same name and see if "select * from temp_view" can return correct result

**Comment by peter-toth** [comment]
So it looks like the test case caused a bit of confusion here so I've split it into 2 cases.
- `SPARK-34421: Resolve temporary objects in temporary views with CTEs` tests the regression part due to https://github.com/apache/spark/pull/30567. Please note that the temp view in temp view part does work even without this fix and the bug affects only the temp function in temp view case. Permanent functions are not affected and the code change modifies `collectTemporaryObjects` only. 
- `SPARK-34421: Resolve temporary objects in permanent views with CTEs` tests the non regression part. That "temp object in permanent" check came with https://github.com/apache/spark/pull/30567 but it doesn't work on permanent views with CTEs without this fix.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: This part isn't, please see https://github.com/apache/spark/pull/31550#issuecomment-781229308 for details.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: While I agree with you, the check came with https://github.com/apache/spark/pull/30567 that is a 3.1 fix only.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: No, permanent functions are not affected: https://github.com/apache/spark/pull/31550#issuecomment-781229308

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Sure, fixed in: https://github.com/apache/spark/pull/31550/commits/ce34b5ee88ac066270a9c6ac6c50cd546ba38929

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala
Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala:560]: as @cloud-fan explained

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala
Comment (linhongliu-db) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: ignore my comment, I found that is not necessary after checked your code one more time.

**Comment by linhongliu-db** [comment]
> SPARK-34421: Resolve temporary objects in permanent views with CTEs tests the non regression part. That "temp object in permanent one" check came with #30567 but it doesn't work on permanent views with CTEs without this fix.

Just a correction, temporary objects check for permanent views are existing for a long time, not introduced by #30567. But we never check the CTE scenario which is fixed by this PR. You can try the below queries on 3.0
```
CREATE TEMPORARY VIEW v0 AS SELECT 0
CREATE VIEW v1 AS SELECT FROM v0
```

**Comment by peter-toth** [comment]
> > [SPARK-34421](https://issues.apache.org/jira/browse/SPARK-34421): Resolve temporary objects in permanent views with CTEs tests the non regression part. That "temp object in permanent one" check came with #30567 but it doesn't work on permanent views with CTEs without this fix.
> 
> Just a correction, temporary objects check for permanent views are existing for a long time, not introduced by #30567. But we never check the CTE scenario which is fixed by this PR. You can try the below queries on 3.0
> 
> ```
> CREATE TEMPORARY VIEW v0 AS SELECT 0
> CREATE VIEW v1 AS SELECT FROM v0
> ```

Indeed, thanks for the correction.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3985]: Can we move the test to `SQLViewTestSuite`, so that it's tested against temp view, global temp view, and permanent view automatically?

**Comment by cloud-fan** [comment]
> Please note that the temp view in temp view part does work even without this fix 

hmm why is that? Looking at the code I think temp view and function should both trigger this bug.

**Comment by cloud-fan** [comment]
> Please note that the temp view in temp view part does work even without this fix 

hmm why is that? Looking at the code I think temp view and function should both trigger this bug.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3967]: What's the behavior before? Fail to read the `test_view`?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala
Comment (linhongliu-db) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3985]: because temp view and permanent view have different behaviors, it's not easy to add this to `SQLViewTestSuite`

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3985]: ah I see, makes sense.

**Comment by cloud-fan** [comment]
BTW let's also put the error message caused by this bug in the PR description, so that people can easily understand what happens, instead of just seeing `a bug that prevents analysis`.

**Comment by peter-toth** [comment]
> BTW let's also put the error message caused by this bug in the PR description, so that people can easily understand what happens, instead of just seeing `a bug that prevents analysis`.

Sure, added to the PR description.

> > Please note that the temp view in temp view part does work even without this fix
> 
> hmm why is that? Looking at the code I think temp view and function should both trigger this bug.

No idea yet, but can take a look at it tomorrow.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
> No idea yet, but can take a look at it tomorrow.

It will be great to figure this out, in case there are other bugs. It doesn't block this PR though.

**Comment by linhongliu-db** [comment]
take a look at "why temp view refers temp view in CTE is working without this PR".
It's because we do `ResolveTempViews` both in `ResolveRelations` and `ResolveTables`.  In this case, `ResolveRelations` try to resolve `temp_view` as `default.temp_view` since it's not in the referred list (caused by this cte bug). But in `ResolveTables`, it just try to resolve all `UnreolvedRelation`s as a temporary view, so in this case, it succeeds (by mistake).

Here is an example to explain:
```sql
use default;
create table same_name as select 100 as a;
create temp view test_view as select * from same_name;
select * from test_view; -- output 100
create temp view same_name as select 99 as a; -- a temp view has the same name with the table
select * from test_view; -- output 100, expected, temp view should not override table
drop table default.same_name; -- drop the table
select * from test_view; -- output 99, wrong result, we expect table not found because we dropped it
```

**Comment by cloud-fan** [comment]
I created `test_view` as a permanent view and tried this example in Spark 3.0, it also gives the wrong result. I think it's another bug of the permanent view that gets inherited by the new SQL temp view. @linhongliu-db can you help to fix it?

**Comment by cloud-fan** [comment]
thanks, merging to master/3.1!

**Comment by cloud-fan** [comment]
@peter-toth can you open backport PRs for 3.0/2.4 to fix the permanent view bug? thanks!

**Comment by peter-toth** [comment]
Thanks for the review. I will open backport PRs soon.

**Comment by peter-toth** [comment]
@cloud-fan, I've opened https://github.com/apache/spark/pull/31592 to backport the change to 3.0.
But it looks like https://issues.apache.org/jira/browse/SPARK-29630 is missing from 2.4 and probably we should backport that change before this one. So shall I open a SPARK-29630 backport PR to 2.4?

**Comment by linhongliu-db** [comment]
> So shall I open a SPARK-29630 backport PR to 2.4?

No need, the function is 2.4 is `verifyTemporaryObjectsNotExists`, just apply your changes to that function should be fine.

**Comment by cloud-fan** [comment]
If we decided to not backport SPARK-29630 to 2.4 at that time, we probably shouldn't backport this fix as well. 3.0 should be good enough.

**Comment by peter-toth** [comment]
> > So shall I open a [SPARK-29630](https://issues.apache.org/jira/browse/SPARK-29630) backport PR to 2.4?
> 
> @peter-toth No need, the function is 2.4 is `verifyTemporaryObjectsNotExists`, just manually apply your changes to that function should be fine.

Yes, found that, but the recursive `verify()` inside `verifyTemporaryObjectsNotExists()` is missing due to missing SPARK-29630 and without that function traversing the inner children would be an ugly change.

> If we decided to not backport [SPARK-29630](https://issues.apache.org/jira/browse/SPARK-29630) to 2.4 at that time, we probably shouldn't backport this fix as well. 3.0 should be good enough.

Ok.

**Pull request closed without merging**
